Src/HALALMock/Services/Communication/Ethernet/TCP/Socket.cpp

```diff
@@ -101,185 +177,104 @@ void Socket::reset(){
+            return;
+        }
+        tx_packet_buffer.pop();
+    }
 }
 
+void Socket::start_receiving(){
+	is_receiving = true;
+    receiving_thread = std::thread(&Socket::receive, this);
+}
+void Socket::stop_receiving() {
+    if (is_receiving) {
+        is_receiving = false;
+        if (receiving_thread.joinable()) {
+            receiving_thread.join(); // wait until the thread finish
+        }
+    }
+}
+void Socket::receive() {
+    while (is_receiving) {
+        uint8_t buffer[1024]; // Buffer for the data
+        ssize_t received_bytes = ::recv(socket_fd, buffer, sizeof(buffer), 0);
+        if (received_bytes > 0) {
+            std::vector<uint8_t> packet(buffer, buffer + received_bytes); 
+            {
+                std::lock_guard<std::mutex> lock(mtx); 
+                rx_packet_buffer.push(packet);
+				process_data(); 
+            }
+        } else if (received_bytes < 0) {
+            std::cout << "Error receiving data\n";
+			socket->state = CLOSING;
+            stop_receiving(); 
+			close();
+        }
+    }
+}
 void Socket::process_data(){
+	
 	while(!rx_packet_buffer.empty()){
-		struct pbuf* packet = rx_packet_buffer.front();
-		rx_packet_buffer.pop();
-		uint8_t* new_data = (uint8_t*)(packet->payload);
-		tcp_recved(socket_control_block, packet->tot_len);
+		{
+			std::std::lock_guard<std::mutex> lock(mtx); 
+			vector<uint8_t> packet = rx_packet_buffer.front();
+			rx_packet_buffer.pop();
+		}
+		uint8_t* new_data = (uint8_t*)(packet.data());
 		Order::process_data(this, new_data);
-		pbuf_free(packet);
 	}
 }
 
 bool Socket::add_order_to_queue(Order& order){
 	if(state == Socket::SocketState::CONNECTED){
 		return false;
 	}
-	struct memp* next_memory_pointer_in_packet_buffer_pool = (*(memp_pools[PBUF_POOL_MEMORY_DESC_POSITION]->tab))->next;
-	if(next_memory_pointer_in_packet_buffer_pool == nullptr){
-		memp_free_pool(memp_pools[PBUF_POOL_MEMORY_DESC_POSITION], next_memory_pointer_in_packet_buffer_pool);
-		return false;
-	}
-
-	uint8_t* order_buffer = order.build();
-
-	struct pbuf* packet = pbuf_alloc(PBUF_TRANSPORT, order.get_size(), PBUF_POOL);
-	pbuf_take(packet, order_buffer, order.get_size());
-	Socket::tx_packet_buffer.push(packet);
+    std::vector<uint8_t> order_buffer = order.build();
+    if (order_buffer.empty()) {
+        std::cout << "Error: building de order buffer\n";
+        return false; 
+    }
+    std::vector<uint8_t> packet(order_buffer); 
+    {
+        std::lock_guard<std::mutex> lock(mutex); 
+        tx_packet_buffer.push(packet); 
+    }
 	return true;
 }
 
 bool Socket::is_connected(){
 	return state == Socket::SocketState::CONNECTED;
 }
 
-err_t Socket::connect_callback(void* arg, struct tcp_pcb* client_control_block, err_t error){
-	IPV4 remote_ip;
-	remote_ip.address = client_control_block->remote_ip;
-	EthernetNode remote_node(remote_ip,client_control_block->remote_port);
+err_t Socket::connect_callback(void* arg, struct tcp_pcb* client_control_block, err_t error){}
```

Comment:
noup, you have to delete the callbacks, or change the interface, there is a param that is `struct_tcp_pcb*` that is an LwIp data type

Reply:
Okay I didn't want it to change it yet due to everything is public, I will change it to private in the HALAL and change the interface the problem, will be the one to set the keepalives